staging/src/k8s.io/kubectl/pkg/cmd/util/kubectl_match_version.go

```diff
@@ -43,26 +50,87 @@ type MatchVersionFlags struct {
 	RequireMatchedServerVersion bool
 	checkServerVersion          sync.Once
 	matchesServerVersionErr     error
+
+	clientVersion            apimachineryversion.Info
+	versionSkewWarningWriter io.Writer
 }
 
 var _ genericclioptions.RESTClientGetter = &MatchVersionFlags{}
 
+var majorMinorVersionExpression = regexp.MustCompile(`^v(\d+).(\d+)`)
+
 func (f *MatchVersionFlags) checkMatchingServerVersion() error {
 	f.checkServerVersion.Do(func() {
 		if !f.RequireMatchedServerVersion {
+			f.warnIfUnsupportedVersionSkew()
 			return
 		}
 		discoveryClient, err := f.Delegate.ToDiscoveryClient()
 		if err != nil {
 			f.matchesServerVersionErr = err
 			return
 		}
-		f.matchesServerVersionErr = discovery.MatchesServerVersion(version.Get(), discoveryClient)
+		f.matchesServerVersionErr = discovery.MatchesServerVersion(f.clientVersion, discoveryClient)
 	})
 
 	return f.matchesServerVersionErr
 }
 
+func (f *MatchVersionFlags) warnIfUnsupportedVersionSkew() {
+	if os.Getenv(suppressVersionSkewWarningEnvironmentVariable) == "true" {
+		return
+	}
+
+	writer := f.versionSkewWarningWriter
+	if writer == nil {
+		writer = os.Stderr
+	}
+
+	discoveryClient, err := f.Delegate.ToDiscoveryClient()
+	if err != nil {
+		fmt.Fprintf(writer, "WARNING: version skew could not be checked because an error occurred while getting the discovery client: %v\n", err)
+		return
+	}
+	serverVersion, err := discoveryClient.ServerVersion()
+	if err != nil {
+		fmt.Fprintf(writer, "WARNING: version skew could not be checked because an error occurred while getting the server version: %v\n", err)
+		return
+	}
+	serverMajorVersion, serverMinorVersion, err := parseMajorMinorVersion(*serverVersion)
+	if err != nil {
+		fmt.Fprintf(writer, "WARNING: version skew could not be checked because an error occurred while parsing the server version: %v\n", err)
+		return
+	}
+
+	clientMajorVersion, clientMinorVersion, err := parseMajorMinorVersion(f.clientVersion)
+	if err != nil {
+		fmt.Fprintf(writer, "WARNING: version skew could not be checked because an error occurred while parsing the client version: %v\n", err)
+		return
+	}
+
+	const supportedSkew = 1
+	if clientMajorVersion != serverMajorVersion || clientMinorVersion < serverMinorVersion-supportedSkew || clientMinorVersion > serverMinorVersion+supportedSkew {
+		fmt.Fprintf(writer, "WARNING: version difference between client (%s) and server (%s) exceeds the supported minor version skew of +/-%d\n",
+			f.clientVersion.GitVersion, serverVersion.GitVersion, supportedSkew)
+	}
+}
+
+func parseMajorMinorVersion(v apimachineryversion.Info) (int, int, error) {
+	submatches := majorMinorVersionExpression.FindStringSubmatch(v.GitVersion)
```

Comment:
Why would you read this instead of directly accessing `v.Major` and `v.Minor`?

Reply:
I started to use v.Major and v.Minor, but saw some comments saying those were deprecated and gitVersion should be used (which is not really a git version, but named that way for legacy reasons it says):

https://github.com/kubernetes/kubernetes/blob/9cd906e9322f75ce48b47613dd9e025c2b06c143/staging/src/k8s.io/client-go/pkg/version/base.go#L37-L58

v.Major and v.Minor come from those fields:

https://github.com/kubernetes/kubernetes/blob/9cd906e9322f75ce48b47613dd9e025c2b06c143/staging/src/k8s.io/component-base/version/version.go#L28-L42

On second look, I think maybe I misinterpreted that comment as saying they are actually being deprecated, when all it really is is a TODO comment saying they might be deprecated in the future, but that has not started happening yet.

Do you think I should just go ahead and use Major and Minor?  If someone decides to someday actually deprecate them, they can come and change them here as part of that.

Reply:
I updated the PR to use major/minor... Since it is just a TODO comment I saw, there is no telling if or when those fields will ever be deprecated (I've seen some TODO comments in the code base that are multiple years old). If someone needs to change it in the future, it should not be hard.

But I like just using major, minor directly as it is simpler.